coordinator/internal/orm/bundle.go

```diff
@@ -0,0 +1,229 @@
+	db = db.Limit(limit)
+
+	var bundles []*Bundle
+	if err := db.Find(&bundles).Error; err != nil {
+		return nil, fmt.Errorf("Bundle.GetUnassignedAndBatchesUnreadyBundles error: %w", err)
+	}
+	return bundles, nil
+}
+
+// UpdateBatchProofsStatusByBatchHash updates the status of batch_proofs_status field for a given bundle hash.
+func (o *Bundle) UpdateBatchProofsStatusByBatchHash(ctx context.Context, bundleHash string, status types.BatchProofsStatus) error {
+	db := o.db.WithContext(ctx)
+	db = db.Model(&Bundle{})
+	db = db.Where("hash = ?", bundleHash)
+
+	if err := db.Update("batch_proofs_status", status).Error; err != nil {
+		return fmt.Errorf("Bundle.UpdateBatchProofsStatusByBatchHash error: %w, bundle hash: %v, status: %v", err, bundleHash, status.String())
+	}
+	return nil
+}
+
+// UpdateProvingStatusFailed updates the proving status failed of a bundle.
+func (o *Bundle) UpdateProvingStatusFailed(ctx context.Context, bundleHash string, maxAttempts uint8, dbTX ...*gorm.DB) error {
+	db := o.db
+	if len(dbTX) > 0 && dbTX[0] != nil {
+		db = dbTX[0]
+	}
+	db = db.WithContext(ctx)
+	db = db.Model(&Bundle{})
+	db = db.Where("hash", bundleHash)
+	db = db.Where("total_attempts >= ?", maxAttempts)
+	db = db.Where("proving_status != ?", int(types.ProvingTaskVerified))
+	if err := db.Update("proving_status", int(types.ProvingTaskFailed)).Error; err != nil {
+		return fmt.Errorf("Bundle.UpdateProvingStatus error: %w, bundle hash: %v, status: %v", err, bundleHash, types.ProvingTaskFailed.String())
+	}
+	return nil
+}
+
+// UpdateProofAndProvingStatusByHash updates the bundle proof and proving status by hash.
+func (o *Bundle) UpdateProofAndProvingStatusByHash(ctx context.Context, hash string, proof []byte, provingStatus types.ProvingStatus, proofTimeSec uint64, dbTX ...*gorm.DB) error {
+	db := o.db
+	if len(dbTX) > 0 && dbTX[0] != nil {
+		db = dbTX[0]
+	}
+
+	updateFields := make(map[string]interface{})
+	updateFields["proof"] = proof
+	updateFields["proving_status"] = provingStatus
+	updateFields["proof_time_sec"] = proofTimeSec
+	updateFields["proved_at"] = utils.NowUTC()
+
+	db = db.WithContext(ctx)
+	db = db.Model(&Bundle{})
+	db = db.Where("hash", hash)
+
+	if err := db.Updates(updateFields).Error; err != nil {
+		return fmt.Errorf("Batch.UpdateProofByHash error: %w, batch hash: %v", err, hash)
+	}
+	return nil
+}
+
+// UpdateBundleAttempts atomically increments the attempts count for the earliest available bundle that meets the conditions.
+func (o *Bundle) UpdateBundleAttempts(ctx context.Context, hash string, curActiveAttempts, curTotalAttempts int16) (int64, error) {
+	db := o.db.WithContext(ctx)
+	db = db.Model(&Bundle{})
+	db = db.Where("hash = ?", hash)
+	db = db.Where("active_attempts = ?", curActiveAttempts)
+	db = db.Where("total_attempts = ?", curTotalAttempts)
+	result := db.Updates(map[string]interface{}{
+		"proving_status":  types.ProvingTaskAssigned,
+		"total_attempts":  gorm.Expr("total_attempts + 1"),
```

Comment:
add indices for coordinator's queries. these indices are added, pls add `total_attempts` and `active_attempts` related indices.
```
CREATE INDEX idx_bundle_index_rollup_status ON bundle(index, rollup_status);
CREATE INDEX idx_bundle_hash ON bundle(hash);
CREATE INDEX idx_bundle_hash_proving_status ON bundle(hash, proving_status);
CREATE INDEX idx_bundle_index_desc ON bundle(index DESC);
CREATE INDEX idx_bundle_start_batch_index ON bundle (start_batch_index) WHERE deleted_at IS NULL;
CREATE INDEX idx_bundle_end_batch_index ON bundle (end_batch_index) WHERE deleted_at IS NULL;
```

Reply:
added [174c1f3](https://github.com/scroll-tech/scroll/pull/1399/commits/174c1f32282335862fdc4e45f199e3dc97e406e6)